src/library.js

```diff
@@ -2263,14 +2263,38 @@ mergeInto(LibraryManager.library, {
     return 0;
   },
 
-  // http://pubs.opengroup.org/onlinepubs/000095399/functions/alarm.html
-  alarm__deps: ['raise', '$callUserCallback'],
-  alarm: function(seconds) {
-    setTimeout(function() {
-      callUserCallback(function() {
-        _raise({{{ cDefine('SIGALRM') }}});
-      });
-    }, seconds*1000);
+  $timers: "= {}",
+
+  // Helper function for setitimer that registers timers with the eventloop.
+  // Timer always fire on the main thread, either diretly from JS (here) or
+  // or when the main thread if busy waiting calling _emscripten_yield.
+  _setitimer_js__sig: 'iid',
+  _setitimer_js__proxy: 'sync',
+  _setitimer_js__deps: ['$timers', '$callUserCallback',
+                        '_emscripten_timeout', 'emscripten_get_now'],
+  _setitimer_js: function(which, timeout_ms) {
+#if RUNTIME_DEBUG
+    dbg('setitimer_js ' + which + ' timeout=' + timeout_ms);
+#endif
+    if (timers[which]) {
+      clearTimeout(timers[which].id);
+      delete timers[which];
+    }
+
+    if (timeout_ms) {
```

Comment:
Looks like this does nothing if `timeout_ms` is 0 - is that intentional?

Reply:
I don't see that in https://linux.die.net/man/2/setitimer - what do you mean? (the function gets a pointer which can be null, but that's different than the time in ms being null?)

Reply:
From `man setitimer` on my system:

```
   setitimer()
       The function setitimer() arms or disarms the timer specified by which, by setting the timer
       to  the value specified by new_value.  If old_value is non-NULL, the buffer it points to is
       used to return the previous value of the timer (i.e., the same information that is returned
       by getitimer()).

       If  either field in new_value.it_value is nonzero, then the timer is armed to initially ex‐
       pire at the specified time.  If both fields in new_value.it_value are zero, then the  timer
       is disarmed.
```

A `timeout_ms == 0` is the same as "both fields in new_value.it_value are zero".   If this wasn't the behaviour I don't see any way to disable or stop of a timer while it was still running.

Reply:
I see, thanks... that second paragraph does not appear in the linux docs link from before. Makes sense now.

Reply:
Added some comments to make it more clear